pyro/contrib/gp/parameterized.py

```diff
@@ -152,47 +132,50 @@ def autoguide(self, name, dist_constructor):
-                if arg_name in self._constraints:
-                    # delete its unconstrained parameter
-                    self.set_constraint(arg_name, constraints.real)
-                delattr(self, arg_name)
-
-        # TODO: create a new argument `autoguide_args` to store other args for other
-        # constructors. For example, in LowRankMVN, we need argument `rank`.
-        p = self._buffers[name]
+                delattr(self, "{}_{}".format(name, arg))
+
+        p = self._priors[name]()  # init_to_sample strategy
         if dist_constructor is dist.Delta:
-            p_map = Parameter(p.detach())
-            self.register_parameter("{}_map".format(name), p_map)
-            self.set_constraint("{}_map".format(name), _get_independent_support(self._priors[name]))
-            dist_args = {"map"}
+            support = _get_independent_support(self._priors[name])
+            if support is constraints.real:
+                p_map = Parameter(p.detach())
+            else:
+                p_map = PyroParam(p.detach(), support)
+            setattr(self, "{}_map".format(name), p_map)
+            dist_args = ("map",)
         elif dist_constructor is dist.Normal:
             loc = Parameter(biject_to(self._priors[name].support).inv(p).detach())
-            scale = Parameter(loc.new_ones(loc.shape))
-            self.register_parameter("{}_loc".format(name), loc)
-            self.register_parameter("{}_scale".format(name), scale)
-            dist_args = {"loc", "scale"}
+            scale = PyroParam(loc.new_ones(loc.shape), constraints.positive)
+            setattr(self, "{}_loc".format(name), loc)
+            setattr(self, "{}_scale".format(name), scale)
+            dist_args = ("loc", "scale")
         elif dist_constructor is dist.MultivariateNormal:
             loc = Parameter(biject_to(self._priors[name].support).inv(p).detach())
             identity = eye_like(loc, loc.size(-1))
-            scale_tril = Parameter(identity.repeat(loc.shape[:-1] + (1, 1)))
-            self.register_parameter("{}_loc".format(name), loc)
-            self.register_parameter("{}_scale_tril".format(name), scale_tril)
-            dist_args = {"loc", "scale_tril"}
+            scale_tril = PyroParam(identity.repeat(loc.shape[:-1] + (1, 1)),
+                                   constraints.lower_cholesky)
+            setattr(self, "{}_loc".format(name), loc)
+            setattr(self, "{}_scale_tril".format(name), scale_tril)
+            dist_args = ("loc", "scale_tril")
         else:
             raise NotImplementedError
 
-        if dist_constructor is not dist.Delta:
-            # each arg has a constraint, so we set constraints for them
-            for arg in dist_args:
-                self.set_constraint("{}_{}".format(name, arg),
-                                    dist_constructor.arg_constraints[arg])
         self._guides[name] = (dist_constructor, dist_args)
 
+    @pyro_method
+    def load_pyro_samples(self):
```

Comment:
Interesting, if this has other uses, we should consider moving it upstream, maybe to
```py
# in pyro/nn/module.py
@pyro_method  # _method is now a misnomer but behavior works :-)
def sample(module):
    assert isinstance(module, PyroModule)
    for submod in module.modules():
        ...
```
Can you make this private, `_load_pyro_samples()` to leave us room for future refactoring?

Reply:
done